### ISSUE TITLE
Fibers remain unassigned in BAD_SKY regions

### ISSUE THREAD

**Issue by Srheft** [desihub/fiberassign#187]
@sbailey, @tskisner:

Points shown in black are BAD_SKY (68719476736) locations. Was the motivation for including them in the sky catalog to have them as the last resort for fibers to get assigned to " made up yet sensible coordinates"?  If yes, then they are not getting picked up by any fibers. 
 

![image](https://user-images.githubusercontent.com/32471902/53541596-cb9db200-3ad7-11e9-95ee-bf6732e69e32.png)

**Comment by sbailey** [comment]
@Srheft please post what input catalogs, `fba_run` command you ran, etc. so that others can reproduce this.

in `fiberassign.targets.get_sciencemask()`, `BAD_SKY` is included as part of the science targets, which should cause them to be placed at low priority.  They are also included in `fiberassign.targets.get_safemask()`, though I thought that functionality was turned off for now.

Issue #175 is related, covering how we want to treat BAD_SKY differently (not just low-priority science targets), but it is worth digging into this case more because it implies that targets in the sciencemask are still being ignored during assignment.  Why?

**Comment by Srheft** [comment]
@sbailey 

- The MTL file is posted [here](http://www.astro.utah.edu/~u6022465/cmx/all_STDtest_SV0bright/).
- The sky catalog is the one from the main survey (skies-dr7.1-PR459.fits), also posted.
- No safe_mask was additionally enforced in my run as a flag as I thought BAD_SKY is the [default](https://github.com/desihub/fiberassign/blob/dcdf482c395ff000fdb5ed1ad439457e1bbbc9ef/py/fiberassign/targets.py#L80)
- The comand was: `fba_run --targets ./MTL_allSTARs.fits ./skies-dr7.1-PR459.fits --footprint ./tiles_allSTD.fits --write_all_targets  --dir ./FAout_allSTDs/ --overwrite`
- Tile center is at (135,32) although I have changed this now to avoid those patches as much as possible. Even if these fibers were assigned, it's best use of fibers to avoid large regions of BAD_SKY for cmx where there is no obligation to fully scan a footprint.

**Comment by tskisner** [comment]
I'm investigating this now.  One thing I noticed is that the tile IDs are larger that 99999, which makes sense for specialized tiling.  I had previously used 6 digits for tile IDs, but then switched it to 5 digits for compatibility with the legacy branch.  This is used in the code when generating names of output files, etc.  Any objections to switching back to 6 digits?

**Comment by sbailey** [comment]
OK to switch to 6 digits now that we have a reference tag that is the new code but backwards compatible with the old code.  We will indeed need 6 digits to accommodate cmx and sv.

**Comment by tskisner** [comment]
I tried the latest version of the "driver" branch on these files.  I think the problem is:

1.  As discussed in #175, BAD_SKY targets are currently classified as "science" targets.

2.  The BAD_SKY targets in this dataset are in the sky file, which does not have a NUMOBS_MORE column.  When reading targets, the default of NUMOBS_MORE if it does not exist in the file is zero.  This is fine for sky, since the obs remaining is ignored for sky.  In this case, the BAD_SKY targets are read in as science targets with zero remaining obs.

The real solution is to fix #175, which is next on my list.  Until then, you could move the bad sky targets to the MTL file and give them some large NUMOBS_MORE value.  Let's leave this issue open until #175 is closed and we verify that this dataset works as intended.

**Comment by tskisner** [comment]
Running the branch in the open PR #189 on this data like this:
```
#!/bin/bash

odir=./FAout_allSTDs
tfile=./tiles_allSTD.fits

export DESI_LOGLEVEL=DEBUG
export OMP_NUM_THREADS=6

fba_run \
    --targets ./MTL_allSTARs.fits ./skies-dr7.1-PR459.fits \
    --footprint ${tfile} \
    --write_all_targets \
    --dir ${odir} \
    --overwrite

export OMP_NUM_THREADS=1

mkdir -p ${odir}_plots
rm -f ${odir}_plots/*.pdf

fba_plot_results \
    --petals 5 \
    --footprint ${tfile} \
    --real_shapes \
    --dir $odir \
    --prefix "fiberassign_" \
    --out ${odir}_plots
```
Produces for example on petal 5 of tile 300005 (the 2 unassigned fibers are the ETC fibers):
![sv_issue-187_300005](https://user-images.githubusercontent.com/84221/53923220-5fb5cf00-402c-11e9-948e-5d878ae50c18.png)
However, I do not see the pattern of unassigned BAD_SKY targets in petal 5 of any of the tiles.  Which tile ID is your plot from in this thread?  After #189 is merged, we should confirm that it fixes your issue.

**Comment by Srheft** [comment]
Did you have the BAD_SKY addded to MTL or just as part of the sky catalog?  as mentioned in email thread and in this ticket, BAD_SKYs were not in my MTL in the example I share here. The picture is from tile 300005 shown [here](http://www.astro.utah.edu/~u6022465/cmx/all_STDtest_SV0bright/fiberassign_300005.pdf). I am now adding BAD_SKY to MTL as you suggested to verify that the issue is resolved.

**Comment by tskisner** [comment]
Hello @srheft, sorry for confusion, the code you ran 7 days ago did indeed have the problem that any bad_sky targets in the sky file would be treated as a science target with zero obs remaining.  My plot above was using the code in PR #189 which was just merged to master.  If you use current master, your plot for tile 300005 should hopefully agree with the one I posted.  This is because these bad_sky targets are now a completely separate type of target which is not assigned until the very end, and which has no obs remaining (like sky).

**Comment by Srheft** [comment]
@tskisner  yes, indeed: @sbailey pinged me to safe_mask that got merged minutes ago and that branch fixed the issue. Thanks for quick turn around on this. I'm sure you saw the two fibers that are not assigned in your plot and mine below though [same tile same petal]. 

![image](https://user-images.githubusercontent.com/32471902/53933831-4b3dfa80-405d-11e9-8bf5-190a495381b8.png)

**Comment by tskisner** [comment]
Those are the sky monitor (ETC) fibers, which are only ever assigned to sky targets.  If we want to assign those to bad_sky as well, then we can make that trivial change.

**Comment by Srheft** [comment]
Ah... that's right. Fiber numbers were small so didn't recognize they are >5000. Thanks.

**Comment by tskisner** [comment]
@kdawson1000, I don't understand the word "documented" in your statement.  The sky monitor fibers are in their own HDU in the final output.  The device ID, location, and assigned sky target are in this HDU.  The reason the ETC fibers were not assigned in this example run is that there were no sky fibers reachable by those fibers.

Was your point just that @dkirby or others who are creating the input sky catalog need to make sure to include sky targets reachable by the ETC fibers?  If that is the case, then I suggest opening an issue with the code that generates the catalog and closing this issue.

The only reason to leave this issue open is if we want fiberassign to fall back to assigning BAD_SKY targets to the ETC fibers in the case where there are no sky targets available.  Is that what we want?

**Comment by sbailey** [comment]
I moved this discussion about sky monitor fiber assignment when no SKY are available to #190 so that this ticket doesn't morph into a different topic than the original post.
